From e93b77f155df86c689d7af481734c9e22a798e13 Mon Sep 17 00:00:00 2001
From: Niall Pemberton <niallp@apache.org>
Date: Sun, 31 Jan 2010 00:53:56 +0000
Subject: [PATCH] Port LANG-538 to 2.x branch - need to call getTime() on a
 calendar sometimes to get it in the right state, otherwise the timezone gets
 out of whack.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/lang/branches/LANG_2_X@904941 13f79535-47bb-0310-9956-ffa450edef68
---
 .../apache/commons/lang/time/FastDateFormat.java    |  1 +
 .../commons/lang/time/FastDateFormatTest.java       | 13 +++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/src/main/java/org/apache/commons/lang/time/FastDateFormat.java b/src/main/java/org/apache/commons/lang/time/FastDateFormat.java
index d2a0f3c0255..1c4b92fae17 100644
--- a/src/main/java/org/apache/commons/lang/time/FastDateFormat.java
+++ b/src/main/java/org/apache/commons/lang/time/FastDateFormat.java
@@ -867,6 +867,7 @@ public StringBuffer format(Date date, StringBuffer buf) {
      */
     public StringBuffer format(Calendar calendar, StringBuffer buf) {
         if (mTimeZoneForced) {
+            calendar.getTime(); /// LANG-538
             calendar = (Calendar) calendar.clone();
             calendar.setTimeZone(mTimeZone);
         }
diff --git a/src/test/java/org/apache/commons/lang/time/FastDateFormatTest.java b/src/test/java/org/apache/commons/lang/time/FastDateFormatTest.java
index 878dd343f14..5bad6e06f2c 100644
--- a/src/test/java/org/apache/commons/lang/time/FastDateFormatTest.java
+++ b/src/test/java/org/apache/commons/lang/time/FastDateFormatTest.java
@@ -331,4 +331,17 @@ public void testLang303() {
         format = (FastDateFormat) SerializationUtils.deserialize( SerializationUtils.serialize( format ) );
         assertEquals(output, format.format(cal));
     }
+
+    public void testLang538() {
+        final String dateTime = "2009-10-16T16:42:16.000Z";
+
+        // more commonly constructed with: cal = new GregorianCalendar(2009, 9, 16, 8, 42, 16)
+        // for the unit test to work in any time zone, constructing with GMT-8 rather than default locale time zone
+        GregorianCalendar cal = new GregorianCalendar(TimeZone.getTimeZone("GMT-8"));
+        cal.clear();
+        cal.set(2009, 9, 16, 8, 42, 16);
+
+        FastDateFormat format = FastDateFormat.getInstance("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", TimeZone.getTimeZone("GMT"));
+        assertEquals("dateTime", dateTime, format.format(cal));
+    }
 }
